### DISCUSSION TITLE
03_pytorch_computer_vision_video

### DISCUSSION THREAD

**Discussion by deordered** [mrdbourke/pytorch-deep-learning#113] [Q&A] [answered]
While trying to functionizing the train and test steps, when trying to measure the time i receive this error on the accuracy_fn!

RuntimeError: Inference tensors cannot be saved for backward. To work around you can make a clone to get a normal tensor and use it in autograd.

**Comment by mrdbourke** [comment, marked as answer]
Hi @deordered,

Do you mind sharing your code (share as much as possible, as in, the code and error too)?

Where about in the sections are you receiving this error?

You can see each of the sections here: https://www.learnpytorch.io/03_pytorch_computer_vision/

**Reply by deordered** [reply]
![Screenshot 2022-08-24 170815](https://user-images.githubusercontent.com/99051310/186440396-115ffd9a-ac70-4fcc-be05-d50b2ecec35a.png)
![Screenshot 2022-08-24 170847](https://user-images.githubusercontent.com/99051310/186440406-ab2fdc42-0561-4444-ac6b-38c8ca62f78c.png)

**Reply by deordered** [reply]
i was in doubt and ran the code!

**Reply by mrdbourke** [reply]
Looks like your error is within the `train_step()`, perhaps check the function and see whether or not you're using `with torch.inference_mode():` (this should only be for `test_step()`).

**Reply by deordered** [reply]
I've rewritten the code & this time I've received the same error but on the (device=device) rather than the (accuracy _fn=accuracy _fn)!

**Reply by deordered** [reply]
@mrdbourke could not find a soltution yet! any hint/help?

**Reply by deordered** [reply]
after many attempts, it's asking me to make a clone to get a normal tensor!
RuntimeError: Inference tensors cannot be saved for backward. To work around you can make a clone to get a normal tensor and use it in autograd.

the code:

```python
torch.manual_seed(42)

# Measure time
from timeit import default_timer as timer
train_time_start_on_gpu = timer()

# Set epochs
epochs = 3

# Create an optimization and evaluation loop using train and test step
for epoch in tqdm(range(epochs)):
  print(f"Epoch: {epoch}\-----")
  train_step(model=model_1,
             data_loader=train_dataloader,
             loss_fn=loss_fn,
             optimizer=optimizer,
             accuracy_fn=accuracy_fn,
             device=device)
  
  test_step(model=model_1,
            data_loader=test_dataloader,
            loss_fn=loss_fn,
            accuracy_fn=accuracy_fn,
            device=device)  # the error is here!
  
train_time_end_on_gpu = timer()

total_train_time_model_1 = print_train_time(start=train_time_start_on_gpu,
                                            end=train_time_end_on_gpu,
                                            device=device)
```

**Reply by mrdbourke** [reply]
Hi @deordered, 

I've just re-ran the code from the notebook and it is working for me.

My guess is that your error is somewhere in these functions: 

```python
def train_step(model: torch.nn.Module,
               data_loader: torch.utils.data.DataLoader,
               loss_fn: torch.nn.Module,
               optimizer: torch.optim.Optimizer,
               accuracy_fn,
               device: torch.device = device):
    train_loss, train_acc = 0, 0
    for batch, (X, y) in enumerate(data_loader):
        # Send data to GPU
        X, y = X.to(device), y.to(device)

        # 1. Forward pass
        y_pred = model(X)

        # 2. Calculate loss
        loss = loss_fn(y_pred, y)
        train_loss += loss
        train_acc += accuracy_fn(y_true=y,
                                 y_pred=y_pred.argmax(dim=1)) # Go from logits -> pred labels

        # 3. Optimizer zero grad
        optimizer.zero_grad()

        # 4. Loss backward
        loss.backward()

        # 5. Optimizer step
        optimizer.step()

    # Calculate loss and accuracy per epoch and print out what's happening
    train_loss /= len(data_loader)
    train_acc /= len(data_loader)
    print(f"Train loss: {train_loss:.5f} | Train accuracy: {train_acc:.2f}%")

def test_step(data_loader: torch.utils.data.DataLoader,
              model: torch.nn.Module,
              loss_fn: torch.nn.Module,
              accuracy_fn,
              device: torch.device = device):
    test_loss, test_acc = 0, 0
    model.eval() # put model in eval mode
    # Turn on inference context manager
    with torch.inference_mode(): 
        for X, y in data_loader:
            # Send data to GPU
            X, y = X.to(device), y.to(device)
            
            # 1. Forward pass
            test_pred = model(X)
            
            # 2. Calculate loss and accuracy
            test_loss += loss_fn(test_pred, y)
            test_acc += accuracy_fn(y_true=y,
                y_pred=test_pred.argmax(dim=1) # Go from logits -> pred labels
            )
        
        # Adjust metrics and print out
        test_loss /= len(data_loader)
        test_acc /= len(data_loader)
        print(f"Test loss: {test_loss:.5f} | Test accuracy: {test_acc:.2f}%\n")
```

I'd check them out and compare them to your code to see what may be wrong.

And be sure to reference the original notebook here: https://github.com/mrdbourke/pytorch-deep-learning/blob/main/03_pytorch_computer_vision.ipynb

**Reply by deordered** [reply]
I did it couple of times and it seems there is something wrong in the function setup as you mentioned! 
I keep getting this error on the accuracy_fn; 
(RuntimeError: Inference tensors cannot be saved for backward. To work around you can make a clone to get a normal tensor and use it in autograd.) 

It seems like there is something wrong with train-step, searched for too long and could not find a solution!
From what I understand, the error says that inferences/evaluation tensors cannot be saved for backward, which is strange and it is suggesting me to clone them, though I am not sure how to approach that!
Help is needed here!

**Reply by deordered** [reply]
fixed!!

**Reply by mrdbourke** [reply]
Glad to hear!

What did you find the error to be?
